From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lolren <loren.bufanu@gmail.com>
Date: Sun, 23 Aug 2026 15:57:12 +0100
Subject: [PATCH] libipa: camera_sensor_helper: Add Sony IMX371 helper

The IMX371 analogue gain follows 1024 / (1024 - code), and its 10-bit
black level is 0x40. Register a helper so IPAs operate in real gain units
and use the correct black level instead of treating the register code as
a linear gain multiplier.

Signed-off-by: Lolren <loren.bufanu@gmail.com>

diff --git a/src/ipa/libipa/camera_sensor_helper.cpp b/src/ipa/libipa/camera_sensor_helper.cpp
index 82bf255..a738a42 100644
--- a/src/ipa/libipa/camera_sensor_helper.cpp
+++ b/src/ipa/libipa/camera_sensor_helper.cpp
@@ -629,6 +629,18 @@ public:
 };
 REGISTER_CAMERA_SENSOR_HELPER("imx355", CameraSensorHelperImx355)
 
+class CameraSensorHelperImx371 : public CameraSensorHelper
+{
+public:
+	CameraSensorHelperImx371()
+	{
+		/* The 10-bit black level is 0x40. */
+		blackLevel_ = 4096;
+		gain_ = AnalogueGainLinear{ 0, 1024, -1, 1024 };
+	}
+};
+REGISTER_CAMERA_SENSOR_HELPER("imx371", CameraSensorHelperImx371)
+
 class CameraSensorHelperImx415 : public CameraSensorHelper
 {
 public:
-- 
2.49.0

